File: app/src/main/java/com/envanto/barcode/utils/LoginManager.java
package com.envanto.barcode.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.envanto.barcode.api.ApiClient;
import com.envanto.barcode.api.LoginResponse;
import com.envanto.barcode.utils.AppConstants;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Kullanıcı girişi için ortak utility sınıfı
 * DeviceAuthManager benzeri modüler yapı
 */
public class LoginManager {
    private static final String TAG = "LoginManager";
    private static final String PREF_NAME = "envanto_login_prefs";
    private static final String SAVED_USERS_PREF = "envanto_saved_users";
    private static final String KEY_USER_TOKEN = "user_token";
    private static final String KEY_USER_EMAIL = "user_email";
    private static final String KEY_USER_NAME = "user_name";
    private static final String KEY_USER_SURNAME = "user_surname";
    private static final String KEY_USER_PERMISSION = "user_permission";
    private static final String KEY_USER_ID = "user_id";
    private static final String KEY_LOGIN_TIME = "login_time";
    private static final String KEY_IS_LOGGED_IN = "is_logged_in";
    private static final String KEY_REMEMBER_ME = "remember_me";
    
    /**
     * Kayıtlı kullanıcı bilgileri için model sınıfı
     */
    public static class SavedUser {
        private String email;
        private String name;
        private String surname;
        private String password; // Encrypted olarak saklanacak
        private int userId;
        private int permission;
        private long lastLoginTime;
        
        public SavedUser() {}
        
        public SavedUser(String email, String name, String surname, String password, int userId, int permission) {
            this.email = email;
            this.name = name;
            this.surname = surname;
            this.password = password;
            this.userId = userId;
            this.permission = permission;
            this.lastLoginTime = System.currentTimeMillis();
        }
        
        // Getters and Setters
        public String getEmail() { return email; }
        public void setEmail(String email) { this.email = email; }
        
        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        
        public String getSurname() { return surname; }
        public void setSurname(String surname) { this.surname = surname; }
        
        public String getPassword() { return password; }
        public void setPassword(String password) { this.password = password; }
        
        public long getLastLoginTime() { return lastLoginTime; }
        public void setLastLoginTime(long lastLoginTime) { this.lastLoginTime = lastLoginTime; }
        
        public int getUserId() { return userId; }
        public void setUserId(int userId) { this.userId = userId; }
        
        public int getPermission() { return permission; }
        public void setPermission(int permission) { this.permission = permission; }
        
        public String getFullName() {
            if (name != null && surname != null) {
                return name + " " + surname;
            }
            return email; // Fallback
        }
    }
    
    /**
     * Kullanıcı girişi durumları için callback interface
     */
    public interface LoginCallback {
        void onLoginSuccess(LoginResponse.User user);
        void onLoginFailure(String message, String errorCode);
        void onShowLoading();
        void onHideLoading();
    }
    
    /**
     * Kullanıcı girişi yapar
     * 
     * @param activity Context olarak kullanılacak AppCompatActivity
     * @param email Kullanıcı e-posta adresi
     * @param password Kullanıcı şifresi
     * @param callback Giriş durumları için callback
     */
    public static void login(AppCompatActivity activity, String email, String password, LoginCallback callback) {
        try {
            // Yükleme ekranını göster
            callback.onShowLoading();
            
            // Input validation
            if (TextUtils.isEmpty(email) || TextUtils.isEmpty(password)) {
                callback.onHideLoading();
                callback.onLoginFailure("E-posta ve şifre alanları boş olamaz", "EMPTY_FIELDS");
                return;
            }
            
            Log.d(TAG, "Kullanıcı girişi başlatılıyor: " + email);
            
            // API çağrısı
            Log.d(TAG, "API çağrısı başlatılıyor - Email: " + email);
            ApiClient.getInstance().getApiService().login("login", email, password)
                .enqueue(new Callback<LoginResponse>() {
                    @Override
                    public void onResponse(Call<LoginResponse> call, Response<LoginResponse> response) {
                        // Kısa bir süre bekle
                        new Handler(Looper.getMainLooper()).postDelayed(() -> {
                            callback.onHideLoading();
                        }, AppConstants.Timing.LOADING_DIALOG_DELAY);
                        
                        if (response.isSuccessful() && response.body() != null) {
                            LoginResponse loginResponse = response.body();
                            
                            if (loginResponse.isSuccess()) {
                                // Giriş başarılı
                                LoginResponse.User user = loginResponse.getUser();
                                
                                // Kullanıcı bilgilerini kaydet
                                saveUserSession(activity, user);
                                
                                Log.d(TAG, "Giriş başarılı: " + user.getFullName());
                                Log.d(TAG, "Kullanıcı yetkisi: " + user.getPermission());
                                
                                // Başarı callback'i çağır
                                callback.onLoginSuccess(user);
                            } else {
                                // Giriş başarısız
                                Log.w(TAG, "Giriş başarısız: " + loginResponse.getMessage());
                                callback.onLoginFailure(loginResponse.getMessage(), loginResponse.getErrorCode());
                            }
                        } else {
                            // Sunucu hatası
                            String errorMsg = "Hata kodu: " + response.code();
                            try {
                                if (response.errorBody() != null) {
                                    errorMsg += " - " + response.errorBody().string();
                                }
                            } catch (Exception e) {
                                errorMsg += " - Hata detayı okunamadı";
                            }
                            Log.e(TAG, "Sunucu yanıt hatası: " + errorMsg);
                            callback.onLoginFailure("Sunucu ile iletişim kurulamadı. Hata: " + response.code(), "SERVER_ERROR");
                        }
                    }
                    
                    @Override
                    public void onFailure(Call<LoginResponse> call, Throwable t) {
                        callback.onHideLoading();
                        
                        // Ağ hatası
                        Log.e(TAG, "Giriş isteği başarısız: " + t.getClass().getSimpleName() + " - " + t.getMessage());
                        Log.e(TAG, "URL: " + call.request().url());
                        callback.onLoginFailure("İnternet bağlantınızı kontrol edin ve tekrar deneyin. Hata: " + t.getMessage(), "NETWORK_ERROR");
                    }
                });
                
        } catch (Exception e) {
            Log.e(TAG, "LoginManager genel hatası: " + e.getMessage());
            callback.onHideLoading();
            callback.onLoginFailure("Beklenmeyen bir hata oluştu.", "UNKNOWN_ERROR");
        }
    }
    
    /**
     * Kullanıcı oturumunu kaydeder
     */
    private static void saveUserSession(Context context, LoginResponse.User user) {
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        
        editor.putBoolean(KEY_IS_LOGGED_IN, true);
        editor.putString(KEY_USER_TOKEN, user.getToken());
        editor.putString(KEY_USER_EMAIL, user.getEmail());
        editor.putString(KEY_USER_NAME, user.getName());
        editor.putString(KEY_USER_SURNAME, user.getSurname());
        editor.putInt(KEY_USER_PERMISSION, user.getPermission());
        editor.putInt(KEY_USER_ID, user.getUserId());
        editor.putLong(KEY_LOGIN_TIME, System.currentTimeMillis());
        
        editor.apply();
        
        Log.d(TAG, "Kullanıcı oturumu kaydedildi");
    }
    
    /**
     * Kullanıcıyı "Beni Hatırla" listesine ekler
     */
    public static void saveUserForRemembering(Context context, String email, String name, String surname, String password, int userId, int permission) {
        List<SavedUser> savedUsers = getSavedUsers(context);
        
        // Aynı e-posta adresi varsa güncelle
        boolean found = false;
        for (SavedUser user : savedUsers) {
            if (user.getEmail().equals(email)) {
                user.setName(name);
                user.setSurname(surname);
                user.setPassword(password);
                user.setUserId(userId);
                user.setPermission(permission);
                user.setLastLoginTime(System.currentTimeMillis());
                found = true;
                break;
            }
        }
        
        // Yeni kullanıcı ekle
        if (!found) {
            savedUsers.add(new SavedUser(email, name, surname, password, userId, permission));
        }
        
        // Maksimum 5 kullanıcı sakla (en son girenleri)
        if (savedUsers.size() > 5) {
            savedUsers.sort((u1, u2) -> Long.compare(u2.getLastLoginTime(), u1.getLastLoginTime()));
            savedUsers = savedUsers.subList(0, 5);
        }
        
        // SharedPreferences'a kaydet
        SharedPreferences prefs = context.getSharedPreferences(SAVED_USERS_PREF, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        
        Gson gson = new Gson();
        String json = gson.toJson(savedUsers);
        editor.putString("saved_users_list", json);
        editor.apply();
        
        Log.d(TAG, "Kullanıcı hatırlanacaklar listesine eklendi: " + email);
    }
    
    /**
     * Kayıtlı kullanıcıları getirir
     */
    public static List<SavedUser> getSavedUsers(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(SAVED_USERS_PREF, Context.MODE_PRIVATE);
        String json = prefs.getString("saved_users_list", "");
        
        if (json.isEmpty()) {
            return new ArrayList<>();
        }
        
        try {
            Gson gson = new Gson();
            Type listType = new TypeToken<List<SavedUser>>(){}.getType();
            List<SavedUser> savedUsers = gson.fromJson(json, listType);
            
            // Null kontrolü
            if (savedUsers == null) {
                return new ArrayList<>();
            }
            
            // Son giriş zamanına göre sırala
            savedUsers.sort((u1, u2) -> Long.compare(u2.getLastLoginTime(), u1.getLastLoginTime()));
            
            return savedUsers;
        } catch (Exception e) {
            Log.e(TAG, "Kayıtlı kullanıcılar yüklenirken hata: " + e.getMessage());
            return new ArrayList<>();
        }
    }
    
    /**
     * Kayıtlı kullanıcıyı siler
     */
    public static void removeSavedUser(Context context, String email) {
        List<SavedUser> savedUsers = getSavedUsers(context);
        savedUsers.removeIf(user -> user.getEmail().equals(email));
        
        // Güncellenmiş listeyi kaydet
        SharedPreferences prefs = context.getSharedPreferences(SAVED_USERS_PREF, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        
        Gson gson = new Gson();
        String json = gson.toJson(savedUsers);
        editor.putString("saved_users_list", json);
        editor.apply();
        
        Log.d(TAG, "Kayıtlı kullanıcı silindi: " + email);
    }
    
    /**
     * Kayıtlı kullanıcı ile otomatik giriş yapar
     */
    public static void loginWithSavedUser(AppCompatActivity activity, SavedUser savedUser, LoginCallback callback) {
        login(activity, savedUser.getEmail(), savedUser.getPassword(), new LoginCallback() {
            @Override
            public void onLoginSuccess(LoginResponse.User user) {
                // Giriş başarılı olduğunda kullanıcıyı güncelle
                saveUserForRemembering(activity, user.getEmail(), user.getName(), user.getSurname(), 
                        savedUser.getPassword(), user.getUserId(), user.getPermission());
                callback.onLoginSuccess(user);
            }
            
            @Override
            public void onLoginFailure(String message, String errorCode) {
                callback.onLoginFailure(message, errorCode);
            }
            
            @Override
            public void onShowLoading() {
                callback.onShowLoading();
            }
            
            @Override
            public void onHideLoading() {
                callback.onHideLoading();
            }
        });
    }
    
    /**
     * Kullanıcının giriş yapıp yapmadığını kontrol eder
     */
    public static boolean isLoggedIn(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return prefs.getBoolean(KEY_IS_LOGGED_IN, false);
    }
    
    /**
     * Kullanıcı oturumunu sonlandırır
     */
    public static void logout(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();
        editor.clear();
        editor.apply();
        
        Log.d(TAG, "Kullanıcı oturumu sonlandırıldı");
    }
    
    /**
     * Giriş yapmış kullanıcının bilgilerini döndürür
     */
    public static LoginResponse.User getCurrentUser(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        
        if (!prefs.getBoolean(KEY_IS_LOGGED_IN, false)) {
            return null;
        }
        
        // User nesnesini manuel olarak oluştur
        LoginResponse.User user = new LoginResponse.User();
        
        // Reflection kullanarak private field'lara değer ata
        try {
            java.lang.reflect.Field emailField = LoginResponse.User.class.getDeclaredField("email");
            emailField.setAccessible(true);
            emailField.set(user, prefs.getString(KEY_USER_EMAIL, ""));
            
            java.lang.reflect.Field nameField = LoginResponse.User.class.getDeclaredField("name");
            nameField.setAccessible(true);
            nameField.set(user, prefs.getString(KEY_USER_NAME, ""));
            
            java.lang.reflect.Field surnameField = LoginResponse.User.class.getDeclaredField("surname");
            surnameField.setAccessible(true);
            surnameField.set(user, prefs.getString(KEY_USER_SURNAME, ""));
            
            java.lang.reflect.Field permissionField = LoginResponse.User.class.getDeclaredField("permission");
            permissionField.setAccessible(true);
            permissionField.set(user, prefs.getInt(KEY_USER_PERMISSION, 0));
            
            java.lang.reflect.Field userIdField = LoginResponse.User.class.getDeclaredField("user_id");
            userIdField.setAccessible(true);
            userIdField.set(user, prefs.getInt(KEY_USER_ID, 0));
            
            java.lang.reflect.Field tokenField = LoginResponse.User.class.getDeclaredField("token");
            tokenField.setAccessible(true);
            tokenField.set(user, prefs.getString(KEY_USER_TOKEN, ""));
            
            java.lang.reflect.Field lastLoginField = LoginResponse.User.class.getDeclaredField("last_login");
            lastLoginField.setAccessible(true);
            long loginTime = prefs.getLong(KEY_LOGIN_TIME, 0);
            lastLoginField.set(user, loginTime > 0 ? new java.util.Date(loginTime).toString() : "");
            
        } catch (Exception e) {
            Log.e(TAG, "getCurrentUser reflection hatası: " + e.getMessage());
            return null;
        }
        
        return user;
    }
    
    /**
     * Kullanıcı token'ını döndürür
     */
    public static String getUserToken(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return prefs.getString(KEY_USER_TOKEN, "");
    }
    
    /**
     * Oturum süresinin dolup dolmadığını kontrol eder (24 saat)
     */
    public static boolean isSessionExpired(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        long loginTime = prefs.getLong(KEY_LOGIN_TIME, 0);
        long currentTime = System.currentTimeMillis();
        long sessionDuration = AppConstants.App.SESSION_DURATION;
        
        return (currentTime - loginTime) > sessionDuration;
    }
} 

File: Services/LoginManager.swift
import Foundation
import SwiftUI

// MARK: - LoginManager Callback Protocols
public protocol LoginCallback {
    func onLoginSuccess(user: User)
    func onLoginFailure(message: String, errorCode: String)
    func onShowLoading()
    func onHideLoading()
}

// MARK: - LoginManager
public class LoginManager: ObservableObject {
    public static let shared = LoginManager()
    
    // MARK: - Constants
    private let loginPrefsKey = "envanto_login_prefs"
    private let savedUsersPrefsKey = "envanto_saved_users"
    private let sessionDuration: TimeInterval = 24 * 60 * 60 // 24 saat
    
    // MARK: - UserDefaults Keys
    private let keyUserToken = "user_token"
    private let keyUserEmail = "user_email"
    private let keyUserName = "user_name"
    private let keyUserSurname = "user_surname"
    private let keyUserPermission = "user_permission"
    private let keyUserId = "user_id"
    private let keyLoginTime = "login_time"
    private let keyIsLoggedIn = "is_logged_in"
    private let keyRememberMe = "remember_me"
    
    private init() {}
    
    // MARK: - Login Method
    public static func login(email: String, password: String, callback: LoginCallback) {
        // Input validation
        guard !email.trimmingCharacters(in: .whitespaces).isEmpty,
              !password.trimmingCharacters(in: .whitespaces).isEmpty else {
            callback.onLoginFailure(message: "E-posta ve şifre alanları boş olamaz", errorCode: "EMPTY_FIELDS")
            return
        }
        
        callback.onShowLoading()
        
        print("LoginManager: Kullanıcı girişi başlatılıyor: \(email)")
        
        // API request parameters
        let parameters = [
            "action": "login",
            "email": email.trimmingCharacters(in: .whitespaces),
            "password": password
        ]
        
        // URL oluştur
        guard let baseUrl = UserDefaults.standard.string(forKey: Constants.UserDefaults.apiBaseURL),
              let url = URL(string: "\(baseUrl)/login_api.asp") else {
            callback.onHideLoading()
            callback.onLoginFailure(message: "API URL'i bulunamadı", errorCode: "INVALID_URL")
            return
        }
        
        // HTTP Request oluştur
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        
        // Parameters'ı encode et
        let bodyString = parameters.map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed) ?? "")" }
            .joined(separator: "&")
        request.httpBody = bodyString.data(using: .utf8)
        
        // API çağrısı
        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                callback.onHideLoading()
            }
            
            if let error = error {
                DispatchQueue.main.async {
                    print("LoginManager: Network error - \(error.localizedDescription)")
                    callback.onLoginFailure(message: "İnternet bağlantınızı kontrol edin ve tekrar deneyin. Hata: \(error.localizedDescription)", errorCode: "NETWORK_ERROR")
                }
                return
            }
            
            guard let data = data else {
                DispatchQueue.main.async {
                    callback.onLoginFailure(message: "Sunucudan yanıt alınamadı", errorCode: "NO_DATA")
                }
                return
            }
            
            // Response'u parse et
            do {
                let loginResponse = try JSONDecoder().decode(LoginResponse.self, from: data)
                
                DispatchQueue.main.async {
                    if loginResponse.success, let user = loginResponse.user {
                        // Giriş başarılı
                        print("LoginManager: Giriş başarılı: \(user.fullName)")
                        print("LoginManager: Kullanıcı yetkisi: \(user.permission)")
                        
                        // Kullanıcı oturumunu kaydet
                        LoginManager.saveUserSession(user: user)
                        
                        callback.onLoginSuccess(user: user)
                    } else {
                        // Giriş başarısız
                        print("LoginManager: Giriş başarısız: \(loginResponse.message)")
                        callback.onLoginFailure(message: loginResponse.message, errorCode: loginResponse.errorCode ?? "LOGIN_FAILED")
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    print("LoginManager: JSON parse error - \(error.localizedDescription)")
                    callback.onLoginFailure(message: "Sunucu yanıtı işlenirken hata oluştu", errorCode: "PARSE_ERROR")
                }
            }
        }.resume()
    }
    
    // MARK: - Save User Session
    private static func saveUserSession(user: User) {
        let defaults = UserDefaults.standard
        
        defaults.set(true, forKey: shared.keyIsLoggedIn)
        defaults.set(user.token, forKey: shared.keyUserToken)
        defaults.set(user.email, forKey: shared.keyUserEmail)
        defaults.set(user.name, forKey: shared.keyUserName)
        defaults.set(user.surname, forKey: shared.keyUserSurname)
        defaults.set(user.permission, forKey: shared.keyUserPermission)
        defaults.set(user.userId, forKey: shared.keyUserId)
        defaults.set(Date().timeIntervalSince1970, forKey: shared.keyLoginTime)
        
        print("LoginManager: Kullanıcı oturumu kaydedildi")
    }
    
    // MARK: - Login Status Check
    public static func isLoggedIn() -> Bool {
        return UserDefaults.standard.bool(forKey: shared.keyIsLoggedIn)
    }
    
    // MARK: - Session Expiration Check
    public static func isSessionExpired() -> Bool {
        let loginTime = UserDefaults.standard.double(forKey: shared.keyLoginTime)
        let currentTime = Date().timeIntervalSince1970
        return (currentTime - loginTime) > shared.sessionDuration
    }
    
    // MARK: - Get Current User
    public static func getCurrentUser() -> User? {
        let defaults = UserDefaults.standard
        
        guard defaults.bool(forKey: shared.keyIsLoggedIn) else {
            return nil
        }
        
        let email = defaults.string(forKey: shared.keyUserEmail) ?? ""
        let name = defaults.string(forKey: shared.keyUserName) ?? ""
        let surname = defaults.string(forKey: shared.keyUserSurname) ?? ""
        let permission = defaults.integer(forKey: shared.keyUserPermission)
        let token = defaults.string(forKey: shared.keyUserToken) ?? ""
        let userId = defaults.integer(forKey: shared.keyUserId)
        let loginTime = defaults.double(forKey: shared.keyLoginTime)
        
        let lastLogin = loginTime > 0 ? DateFormatter().string(from: Date(timeIntervalSince1970: loginTime)) : nil
        
        return User(email: email, name: name, surname: surname, permission: permission, token: token, lastLogin: lastLogin, userId: userId)
    }
    
    // MARK: - Get User Token
    public static func getUserToken() -> String {
        return UserDefaults.standard.string(forKey: shared.keyUserToken) ?? ""
    }
    
    // MARK: - Logout
    public static func logout() {
        let defaults = UserDefaults.standard
        
        // Login ile ilgili tüm key'leri temizle
        defaults.removeObject(forKey: shared.keyIsLoggedIn)
        defaults.removeObject(forKey: shared.keyUserToken)
        defaults.removeObject(forKey: shared.keyUserEmail)
        defaults.removeObject(forKey: shared.keyUserName)
        defaults.removeObject(forKey: shared.keyUserSurname)
        defaults.removeObject(forKey: shared.keyUserPermission)
        defaults.removeObject(forKey: shared.keyUserId)
        defaults.removeObject(forKey: shared.keyLoginTime)
        
        print("LoginManager: Kullanıcı oturumu sonlandırıldı")
    }
    
    // MARK: - Saved Users Management
    public static func saveUserForRemembering(email: String, name: String, surname: String, password: String, userId: Int, permission: Int) {
        var savedUsers = getSavedUsers()
        
        // Aynı e-posta adresi varsa güncelle
        if let index = savedUsers.firstIndex(where: { $0.email == email }) {
            savedUsers[index] = SavedUser(email: email, name: name, surname: surname, password: password, userId: userId, permission: permission)
        } else {
            // Yeni kullanıcı ekle
            savedUsers.append(SavedUser(email: email, name: name, surname: surname, password: password, userId: userId, permission: permission))
        }
        
        // Maksimum 5 kullanıcı sakla (en son girenleri)
        if savedUsers.count > 5 {
            savedUsers.sort { $0.lastLoginTime > $1.lastLoginTime }
            savedUsers = Array(savedUsers.prefix(5))
        }
        
        // UserDefaults'a kaydet
        if let encoded = try? JSONEncoder().encode(savedUsers) {
            UserDefaults.standard.set(encoded, forKey: shared.savedUsersPrefsKey)
        }
        
        print("LoginManager: Kullanıcı hatırlanacaklar listesine eklendi: \(email)")
    }
    
    public static func getSavedUsers() -> [SavedUser] {
        guard let data = UserDefaults.standard.data(forKey: shared.savedUsersPrefsKey),
              let savedUsers = try? JSONDecoder().decode([SavedUser].self, from: data) else {
            return []
        }
        
        // Son giriş zamanına göre sırala
        return savedUsers.sorted { $0.lastLoginTime > $1.lastLoginTime }
    }
    
    public static func removeSavedUser(email: String) {
        var savedUsers = getSavedUsers()
        savedUsers.removeAll { $0.email == email }
        
        // Güncellenmiş listeyi kaydet
        if let encoded = try? JSONEncoder().encode(savedUsers) {
            UserDefaults.standard.set(encoded, forKey: shared.savedUsersPrefsKey)
        }
        
        print("LoginManager: Kayıtlı kullanıcı silindi: \(email)")
    }
    
    // MARK: - Auto Login with Saved User
    public static func loginWithSavedUser(savedUser: SavedUser, callback: LoginCallback) {
        login(email: savedUser.email, password: savedUser.password, callback: LoginAutoUpdateCallback(savedUser: savedUser, originalCallback: callback))
    }
}

// MARK: - Helper Callback for Auto Login
private class LoginAutoUpdateCallback: LoginCallback {
    private let savedUser: SavedUser
    private let originalCallback: LoginCallback
    
    init(savedUser: SavedUser, originalCallback: LoginCallback) {
        self.savedUser = savedUser
        self.originalCallback = originalCallback
    }
    
    func onLoginSuccess(user: User) {
        // Giriş başarılı olduğunda kullanıcıyı güncelle
        LoginManager.saveUserForRemembering(email: user.email, name: user.name, surname: user.surname, password: savedUser.password, userId: user.userId, permission: user.permission)
        originalCallback.onLoginSuccess(user: user)
    }
    
    func onLoginFailure(message: String, errorCode: String) {
        originalCallback.onLoginFailure(message: message, errorCode: errorCode)
    }
    
    func onShowLoading() {
        originalCallback.onShowLoading()
    }
    
    func onHideLoading() {
        originalCallback.onHideLoading()
    }
} 